### PR TITLE
Added base CLI binary.

### PR DESCRIPTION
**Pull request by tamtamchik**
Added base CLI implementation (#29). But I think better to move it into separate repo, somethig like `jolicode/jolinotif-cli`.

### Installation

Install package globally:

```bash
$ composer global require jolicode/jolinotif
```
**Note!** Make sure to place the `~/.composer/vendor/bin` directory (or the equivalent directory for your OS) in your PATH so the transfer executable can be located by your system. Simply add this directory to your PATH in your `~/.bashrc` (or `~/.bash_profile`) like this:

```
$ echo "export PATH=~/.composer/vendor/bin:$PATH" >> ~/.bashrc
$ source ~/.bashrc
```

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -22,7 +22,8 @@
     },
     "require": {
         "php": ">=5.4.0",
-        "symfony/process": "~2.3|~3.0"
+        "symfony/process": "~2.3|~3.0",
+        "league/climate": "^3.2"
```
Comment (pyrech) [inline composer.json]: I would prefer to not force regular JoliNotif users to depend on this package. So maybe we should move this requirement in the suggest section and to add a check in the binary, something like that:

```php
// ...
require_once('vendor/autoload.php');

if (!class_exist(CLImate::class)) {
    echo 'CLImate is required to use this script, you can install it by requiring `league/climate`;
    exit(1);
}

$cli = new CLImate;
// ...
```

File: .gitignore [hunk 2]
```diff
@@ -12,3 +12,5 @@ bin/php-cs-fixer
 bin/phpunit
 
 .php_cs.cache
+
+/.idea
```
Comment (pyrech) [inline .gitignore]: This should be in your global gitignore on your system :wink:

File: jolinotif [hunk 3]
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/php -q
+<?php
+
+use Joli\JoliNotif\Notification;
+use Joli\JoliNotif\NotifierFactory;
+use League\CLImate\CLImate;
+
+require_once('vendor/autoload.php');
+
+const VERSION = '1.0.0';
```
Comment (pyrech) [inline jolinotif]: I'm not a huge fan to maintain a version number in this script :confused:

**Comment by tamtamchik** [comment]
JFYI, I'm trying to get rid of any extra dependencies now. Soon will update PR...

**Comment by pyrech** [comment]
Hello @tamtamchik did you got time to finish this PR? Thanks :slightly_smiling_face:

**Comment by tamtamchik** [comment]
Thanks for reminding! PR updated.

File: jolinotif [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    protected $rules = [];
+    protected $arguments = [];
+
+    private $command;
+
+    /**
+     * @param string $description
+     * @param array $rules
+     */
+    public function __construct($description, array $rules)
```
[the remark this answers is not in the document]
Comment (lyrixx) [reply jolinotif]: Then rules and the description are hard-coded in this file. I would be simpler to move them as a class constant.

File: jolinotif [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    /**
+     * Return a value of an CLI argument passed.
+     * @param string $name
+     * @return mixed
+     */
+    public function get($name)
+    {
+        return $this->arguments[$name] ?: false;
```
Comment (lyrixx) [inline jolinotif]: `$this->arguments[$name]` can be undefined

File: jolinotif [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    /**
+     * Checks if an CLI argument was passed.
+     * @param string $name
+     * @return mixed
+     */
+    public function defined($name)
```
Comment (lyrixx) [inline jolinotif]: usually, we name this method `has`. But has this class mix many different concern, I would name it `hasOption`. same for get: `getOption`

File: jolinotif [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+     */
+    public function get($name)
+    {
+        return $this->arguments[$name] ?: false;
+    }
+
+    /**
+     * Checks if an CLI argument was passed.
+     * @param string $name
+     * @return mixed
```
Comment (lyrixx) [inline jolinotif]: it returns a boolean. But insteand of fixing it, please remove all useless PHPDoc. Instead use type hint + meaningful method name

File: jolinotif [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+
+    /**
+     * Log messages to CLI output.
+     *
+     * @param integer $level
+     * @param string $message
+     */
+    private function log($level, $message)
+    {
+        $color = "\033[0m";
```
Comment (lyrixx) [inline jolinotif]: this does not work on windows (ie non posix env)

File: jolinotif [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+    'help'     => [
+        'name' => 'help',
+        'info' => 'Show this help.',
+        'flag' => true,
+    ],
+];
+
+$cli = new Cli($description, $rules);
+
+// Trying to parse args.
```
Comment (lyrixx) [inline jolinotif]: all theses docs are useless

**Comment by tamtamchik** [comment]
Wow, @lyrixx thx review! Great feedback!

File: jolinotif (hunk 4 above)
[the remark this answers is not in the document]
Comment (pyrech) [reply jolinotif]: I think it's time to make a JoliNotif release that bumps PHP requirement to 7.0, so I agree to add scalar typehint where needed

[the remark this answers is not in the document]
Comment (lyrixx) [reply jolinotif]: Nice. More over, the previous released will still be download from github for older PHP version

**Comment by tamtamchik** [comment]
@pyrech I'm kind of a bit busy at the moment, but after NY I'll update the PR to match all comments above ;)

**Comment by pyrech** [comment]
Ok, no problem, take your time! Merry Christmas! :slightly_smiling_face:

**Comment by tamtamchik** [comment]
Sorry for the delay. Updated PR.

File: jolinotif [hunk 10]
```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env php
+<?php
+
+use Joli\JoliNotif\Notification;
+use Joli\JoliNotif\NotifierFactory;
+
+require_once('vendor/autoload.php');
```
Comment (lyrixx) [inline jolinotif]: We used to write `require __DIR__.'vendor/autoload.php';`

**Comment by pyrech** [comment]
I think everything looks good here! @tamtamchik Is there anything you want to change or can I merge it now?

**Comment by tamtamchik** [comment]
LGTM. :shipit: 
Thank you @pyrech and @lyrixx for a great feedback!

**Comment by pyrech** [comment]
Awesome! Thank for your work on this @tamtamchik!

**Comment by lyrixx** [comment]
Thank you @tamtamchik

**Comment by pyrech** [comment]
FYI this PR was just released as v1.3.0 :wink:

**Pull request merged**